src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php

```diff
@@ -490,6 +532,42 @@ private function createAuthenticationListeners(ContainerBuilder $container, stri
         return [$listeners, $defaultEntryPoint];
     }
 
+    private function getUserProvider(ContainerBuilder $container, string $id, array $firewall, string $factoryKey, ?string $defaultProvider, array $providerIds, ?string $contextListenerId): ?string
+    {
+        if (isset($firewall[$factoryKey]['provider'])) {
+            if (!isset($providerIds[$normalizedName = str_replace('-', '_', $firewall[$factoryKey]['provider'])])) {
+                throw new InvalidConfigurationException(sprintf('Invalid firewall "%s": user provider "%s" not found.', $id, $firewall[$factoryKey]['provider']));
+            }
+
+            return $providerIds[$normalizedName];
+        }
+
+        if ('remember_me' === $factoryKey || 'anonymous' === $factoryKey) {
+            if ('remember_me' === $factoryKey && $contextListenerId) {
+                $container->getDefinition($contextListenerId)->addTag('security.remember_me_aware', ['id' => $id, 'provider' => 'none']);
+            }
+
+            // RememberMeFactory will use the firewall secret when created
+            return null;
```

Comment:
I mention in one other spot (I think) that I think we can remove this extra hackery (which is not your code, I realize - it's always been there). This return of this method - `getUserProvider()` - is used as an argument to either the factory `create()` (in the old system) or `createAuthenticator()` (in the new system). In the old system, both `RememberMeFactory::create()` and `AnonymousFactory::create()` don't even use their `$userProvider` argument. So what's the harm in returning the same user provider id as if these weren't special? It would simply mean that both factory classes are passed an argument... that they don't already use. Basically, I don't see any point to having this extra weird logic.

Reply:
I've had to revert this change, as tests like https://github.com/wouterj/symfony/blob/security/deprecate-providers-listeners/src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php#L213-L233 were broken.

I've now slightly modified this, to pass the "all user providers" user provider ID for RememberMe and Anonymous if no user provider was explicitly configured. We may want to deprecate this implict behavior and always require a user to configure a user provider in another PR (see also for instance https://github.com/symfony/maker-bundle/pull/546#discussion_r380160396 ).